src/lib/lossless_cast.cpp

```diff
@@ -0,0 +1,37 @@
+#include "lossless_cast.hpp"
+
+#include "resolve_type.hpp"
+
+namespace opossum {
+
+std::optional<AllTypeVariant> lossless_variant_cast(const AllTypeVariant& variant, DataType target_data_type) {
+  const auto source_data_type = data_type_from_all_type_variant(variant);
+
+  // Safe casting from NULL to NULL is always NULL. (Cannot be handled below as resolve_data_type()
+  // doesn't resolve NULL)
+  if (source_data_type == DataType::Null && target_data_type == DataType::Null) {
+    return NullValue{};
+  }
+
+  // Safe casting between NULL and non-NULL type is not possible. (Cannot be handled below as resolve_data_type()
+  // doesn't resolve NULL)
+  if ((source_data_type == DataType::Null) != (target_data_type == DataType::Null)) {
```

Comment:
I feel like the caller should have ruled out NULL before calling this method.

Reply:
na, they haven't. E.g., for a query `... WHERE a > NULL` the call to lossless_variant_cast is precisely where hyrise figures out that NULL has no representation in whatever data type `a` is.